### PR TITLE
Allow ext keys

### PR DESCRIPTION
**Pull request by Eunovo**
- Add support for extended keys to sp()
- Add sp() to wallet_miniscript test
 - this requires that GetNewDestination returns the same address that `rpc: deriveaddress` returns
 - move basic label implementation to separate methods to achieve this (label implementation to be expanded in later PR)
- SP wallet requires scan private key so ensure that private keys are not disabled in sp wallet

### CODE REVIEW

**Comment by Sjors** [comment]
> SP wallet requires scan private key so ensure that private keys are not disabled in sp wallet

Bit of a wallet philosophy thing, but I think a watch key should _not_ count as a private key. It's much more similar in functionality and security than an xpub.

In other words, I think it should be possible to import these into a watch-only wallet.

**Comment by Sjors** [comment]
I did manage to succesfully import into a wallet with `disable_private_keys=false`, though it gave a warning:

```json
[
  {
    "success": true,
    "warnings": [
      "Not all private keys provided. Some wallet functionality may return unexpected errors"
    ]
  }
]
```

This can probably be avoided with the above.

**Comment by Sjors** [comment]
`getnewaddress` produces the same address each time now...

**Comment by Eunovo** [comment]
> `getnewaddress` produces the same address each time now...

@Sjors I'm thinking about using the existing `label` option to decide between returning the default silent payment address or a labelled address. WDYT? 

> Bit of a wallet philosophy thing, but I think a watch key should not count as a private key. It's much more similar in functionality and security than an xpub.
>
> In other words, I think it should be possible to import these into a watch-only wallet.
 
Makes sense. I'll add a special case to `AddWalletDescriptor` for silent payments

**Comment by Sjors** [comment]
> I'm thinking about using the existing label option to decide between returning the default silent payment address or a labelled address. WDYT?

I think that's a good idea; that had me confused as well.

Perhaps when reusing a label, you could return the same address. There's currently no convenient way to get all labels and their addresses - that might require a new RPC.

**Comment by Sjors** [comment]
Perhaps I did something wrong cherry-picking your commits, but trying to import into a watch-only wallet throws an error for me:

```
bitcoin-cli importdescriptors '[{"desc": "sp(sppub1q...)#00000000", "active":true, "timestamp": 17...}]'
error code: -1
error message:
map::at:  key not found
```

It seems to have added the descriptor twice, one marked as `active` and the other not (and with the wrong spub value and timestamp).

It started a rescan but it doesn't seem to make progress.

Calling `rescanblockchain` triggers the above error again, so it probably happens during the rescan phase of importing. It did manage to find a transaction, as evidenced by the log, though it doesn't show up in the wallet.

**Comment by Eunovo** [comment]
@Sjors Most of this stuff is still a work in progress. The idea is to open PRs to josibake/implement-bip352-full with new functionality gradually. I'll investigate these errors that you have pointed out.

**Comment by Eunovo** [comment]
> Perhaps I did something wrong cherry-picking your commits, but trying to import into a watch-only wallet throws an error for me

Did you cherry-pick on top of josibake/implement-bip352-full?

**Comment by Sjors** [comment]
@Eunovo I rebased the top commits from `josibake/implement-bip352-full` onto `josibake/implement-bip352-sending` and then applied the ones in the PR on top of that. I'll try again later when the branches are updated.

**Comment by josibake** [comment]
> Bit of a wallet philosophy thing, but I think a watch key should not count as a private key. It's much more similar in functionality and security than an xpub.

Strongly agree. We even talked about renaming scan key to "scan entropy," since it is _not_ a private key (i.e. it is not used to secure funds). A scan key + spend pubkey (`sppub`) is functionally equivalent to an xpub, so we should be able to use this in a watch-only wallet. Further more, when calling `listdescriptors`, it should show `sp(sppub...)` which is the scan key || spend pubkey, whereas `listdescriptors` with private_keys = true should show `sp(spprv...)` which is the scan key + spend private key.

> Perhaps when reusing a label, you could return the same address (maybe emit a warning).

I would expect `getnewaddress address_type=silent-payment` to always return the un-labeled silent payment address. Calling `getnewaddress address_type=silent-payment label=xyz` should always return the same labeled address (i.e., the labeled address corresponding to xyz).

Also, I saw this comment from @Sjors earlier but can't find it now, but if a sp descriptor is imported into a wallet, this should set the `silent_payments` flag to true.

**Comment by Eunovo** [comment]
@josibake and @Sjors I stopped storing the scan key to DB in https://github.com/josibake/bitcoin/pull/3/commits/8ed833b5d274ad8f445077034fdfe954afee4ce8 since the desc always contains the scan key, we don't need store it separately. The scan key is also no longer added to the private key when the descriptor is parsed so the watch-only wallets should not have a problem with sp descs

**Comment by Sjors** [comment]
I'm still getting "Cannot import private keys to a wallet with private keys disabled" when importing an `sppub` with 87c1b71402b99ab598726ffc0dbbba1ce843c9ba.

**Comment by Eunovo** [comment]
> I'm still getting "Cannot import private keys to a wallet with private keys disabled" when importing an `sppub` with [87c1b71](https://github.com/josibake/bitcoin/commit/87c1b71402b99ab598726ffc0dbbba1ce843c9ba).

Sorry about that slight error in [87c1b71](https://github.com/josibake/bitcoin/commit/87c1b71402b99ab598726ffc0dbbba1ce843c9ba). I've already fixed this. Will push after I get signing test working for sp descriptor in `test/functional/wallet_miniscript.py`

**Comment by Eunovo** [comment]
> Loading a silent payment wallet still crashes with `sp_spk_man != nullptr` in walletdb.cpp, line 962.

I wasn't able to reproduce this. See results below:
```
> dbtcd
Bitcoin Core starting
> dbtcli getblockchaininfo
{
  "chain": "regtest",
  "blocks": 0,
  "headers": 0,
  "bestblockhash": "0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206",
  "difficulty": 4.656542373906925e-10,
  "time": 1296688602,
  "mediantime": 1296688602,
  "verificationprogress": 1,
  "initialblockdownload": true,
  "chainwork": "0000000000000000000000000000000000000000000000000000000000000002",
  "size_on_disk": 293,
  "pruned": false,
  "warnings": [
    "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
  ]
}
> dbtcli loadwallet alice
{
  "name": "alice"
}
> dbtcli listdescriptors
{
  "wallet_name": "alice",
  "descriptors": [
    {
      "desc": "sp(sprtpub1qqqqqqqqqqqqqqz2cezqye6220rn3xzpxu5lxdh8l89x6mvvz25vx699mlz5tss635p7y6txdrl97smz23cmxylj0wmtygzulgc9ynpctsj8g3zxp9eycgcpvngw9)#nnrxeufl",
      "timestamp": 1296688602,
      "active": true,
      "internal": false
    }
  ]
}
```

Note that `dbtcd` and `dbtcli` are aliases for `src/bitcoind` and `src/bitcoin-cli` respectively

File: src/wallet/walletutil.cpp [hunk 1]
```diff
@@ -97,9 +97,9 @@ WalletDescriptor GenerateWalletDescriptor(const CExtKey& master_key, const Outpu
             }
         }
 
-        // Add the derived keys
+        // We only need to save the spend key
+        // The scan key is preserved in the descriptor
```
Comment (josibake) [inline src/wallet/walletutil.cpp:101]: in https://github.com/josibake/bitcoin/pull/3/commits/a065b4b2073b80177a64a7f87b29d97193ab55d7:

What would be the harm in persisting the key to the database? I think this is fine, but I am suspicious there might be unintended consequences from not persisting the key. Will need to study this section a bit more, thought, to understand how descriptors are persisted.

File: src/wallet/scriptpubkeyman.cpp [hunk 2]
```diff
@@ -2876,7 +2876,7 @@ util::Result<CTxDestination> SilentPaymentDescriptorScriptPubKeyMan::GetNewDesti
     return dest;
 }
 
-CTxDestination SilentPaymentDescriptorScriptPubKeyMan::GetLabelledDestination(uint64_t index)
+V0SilentPaymentDestination SilentPaymentDescriptorScriptPubKeyMan::GetLabelledDestination(uint64_t index)
```
Comment (josibake) [inline src/wallet/scriptpubkeyman.cpp]: in https://github.com/josibake/bitcoin/pull/3/commits/817e93128c28b981257181b82a838bc19ff202ad:

why the change to V0SilentPaymentDestination? Seems like this was fine as a CTxDestination?

**Review by josibake — COMMENTED** [review]
LGTM, left some comments/questions but going to go ahead and merge and start testing. Can address any comments in follow-ups

File: src/wallet/walletutil.cpp (hunk 1 above)
Comment (Eunovo) [reply src/wallet/walletutil.cpp:101]: No harm that I can think of, It just requires more code. Currently, the wallet does not allow the persistence of private keys when the private keys are disabled. If we don't persist the scan key, then this error isn't triggered on watch only wallets.

File: src/rpc/output_script.cpp [hunk 3]
```diff
@@ -285,6 +285,10 @@ static RPCHelpMan deriveaddresses()
 
                     addresses.push_back(EncodeDestination(dest));
                 }
+
+                for (const auto& sppubkey_pair: provider.sppubkeys) {
+                    addresses.push_back(EncodeDestination(V0SilentPaymentDestination(sppubkey_pair.second)));
+                }
```
[the remark this answers is not in the document]
Comment (Eunovo) [reply src/rpc/output_script.cpp:291]: This code always runs at least once. See line `268`

File: src/wallet/scriptpubkeyman.cpp (hunk 2 above)
Comment (Eunovo) [reply src/wallet/scriptpubkeyman.cpp]: I reversed the change locally and the compiles successfully. I'm not sure why I changed it initially.

**Comment by Sjors** [comment]
> I wasn't able to reproduce this

I'll let you know if I run into the crash in later updates. Haven't tested again yet.
